synapse/handlers/e2e_keys.py

```diff
@@ -511,6 +513,262 @@ def upload_signing_keys_for_user(self, user_id, keys):
+                                pubkey = devices[signing_device_id]["keys"]["keys"][
+                                    signing_key_id
+                                ]
+                                verify_key = decode_verify_key_bytes(
+                                    signing_key_id, decode_base64(pubkey)
+                                )
+                                device["signatures"] = sigs
+                                try:
+                                    verify_signed_json(device, user_id, verify_key)
+                                except SignatureVerifyException:
+                                    raise SynapseError(
+                                        400,
+                                        "Invalid signature",
+                                        Codes.INVALID_SIGNATURE,
+                                    )
+
+                                master_key_signature_list.append(
+                                    (signing_key_id, user_id, device_id, signature)
+                                )
+
+                            signature_list.extend(master_key_signature_list)
+                            continue
+
+                        # at this point, we have a device that should be signed
+                        # by the self-signing key
+                        if self_signing_key_id not in device["signatures"][user_id]:
+                            # no signature was sent
+                            raise SynapseError(
+                                400, "Invalid signature", Codes.INVALID_SIGNATURE
+                            )
+
+                        stored_device = None
+                        try:
+                            stored_device = devices[device_id]["keys"]
+                        except KeyError:
+                            raise SynapseError(404, "Unknown device", Codes.NOT_FOUND)
+                        if self_signing_key_id in stored_device.get(
+                            "signatures", {}
+                        ).get(user_id, {}):
+                            # we already have a signature on this device, so we
+                            # can skip it, since it should be exactly the same
+                            continue
+
+                        _check_device_signature(
+                            user_id, self_signing_verify_key, device, stored_device
+                        )
+
+                        signature = device["signatures"][user_id][self_signing_key_id]
+                        signature_list.append(
+                            (self_signing_key_id, user_id, device_id, signature)
+                        )
+                    except SynapseError as e:
+                        failures.setdefault(user_id, {})[
+                            device_id
+                        ] = _exception_to_failure(e)
+            except SynapseError as e:
+                failures[user_id] = {
```

Comment:
kinda wondering under what circumstances we are likely to hit this case, and whether it is worth special handling rather than just failing the whole request.

Reply:
I'm not sure.  It's probably a spec discussion about whether the server can fail the whole request, and under what conditions.